File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// TC- O(n)
// SC- O(1)

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        // base case: if no list, or if there is only one node, or if left==right
        if(!head or left==right)
            return head;

        ListNode* dummy = new ListNode(0, head);
        ListNode* prev = dummy;

        // move prev to node before left
        for(int i=1; i<left; i++)
            prev = prev->next;

        // reverse iteratively
        ListNode* curr = prev->next;
        for(int i=0; i<right-left; i++){
            ListNode* temp = curr->next;
            curr->next = temp->next; // remove temp
            temp->next = prev->next; // put temp before front of reversed list
            prev->next = temp;
            
        }

        return dummy->next;
    }
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# TC- O(n)
# SC- O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if head is None or head.next is None or left==right:
            return head

        dummy = ListNode()
        dummy.next = head
        prev = dummy
        for _ in range(left-1):
            prev = prev.next
        
        curr = prev.next
        tail = curr
        for _ in range(right-left+1):
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        
        tail.next = curr
        before = dummy
        for _ in range(left - 1):
            before = before.next
        before.next = prev

        return dummy.next